service/storage_service.cc

```diff
@@ -6155,6 +6191,264 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+    }
+
+    if (params.snitch_name != _db.local().get_snitch_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Snitch name check failed. This node cannot join the cluster "
+                                "because it uses \"{}\" and not \"{}\"",
+                                params.snitch_name,
+                                _db.local().get_snitch_name()),
+        };
+        co_return result;
+    }
+
+    co_await _topology_state_machine.event.when([this] {
+        // The first node defines the cluster and inserts its entry to the
+        // `system.topology` without checking anything. It is unlikely but
+        // possible that the `join_node_request_handler` fires before the first
+        // node inserts its entry, therefore we might need to wait
+        // until that happens, here.
+        return !_topology_state_machine._topology.is_empty();
+    });
+
+    auto& g0_server = _group0->group0_server();
+    if (params.replaced_id && *params.replaced_id == g0_server.current_leader()) {
+        // There is a peculiar case that can happen if the leader is killed
+        // and then replaced very quickly:
+        //
+        // - Cluster with nodes `A`, `B`, `C` - `A` is the topology
+        //   coordinator/group0 leader,
+        // - `A` is killed,
+        // - New node `D` attempts to replace `A` with the same IP as `A`,
+        //   sends `join_node_request` rpc to node `B`,
+        // - Node `B` handles the RPC and wants to perform group0 operation
+        //   and wants to perform a barrier - still thinks that `A`
+        //   is the leader and is alive, sends an RPC to its IP,
+        // - `D` accidentally receives the request that was meant to `A`
+        //   but throws an exception because of host_id mismatch,
+        // - Failure is propagated back to `B`, and then to `A` - and `A`
+        //   fails the replace operation.
+        //
+        // We can try to detect if this failure might happen: if the new node
+        // is going to replace but the ID of the replaced node is the same
+        // as the leader, wait for a short while until a reelection happens.
+        // If replaced ID == leader ID, then this indicates either the situation
+        // above or an operator error (actually trying to replace a live node).
+
+        const auto timeout = std::chrono::seconds(10);
+
+        slogger.warn("raft topology: the node {} which was requested to be"
+                " replaced has the same ID as the current group 0 leader ({});"
+                " this looks like an attempt to join a node with the same IP"
+                " as a leader which might have just crashed; waiting for"
+                " a reelection",
+                params.host_id, g0_server.current_leader());
+
+        abort_source as;
+        timer<lowres_clock> t;
+        t.set_callback([&as] {
+            as.request_abort();
+        });
+        t.arm(timeout);
```

Comment:
There is `with_timeout` in `seastar/core/with_timeout.hh`

Reply:
Personally, I'm not a big fan of `seastar::with_timeout` because it doesn't properly cancel the future that is being waited on. I prefer telling fibers to abort and then waiting until they gracefully finish.

Reply:
You can pass abort source to the function you execute under `with_timeout` and wait through a gate. It may still be simpler than to re-invent it here. But I agree that `with_timeout` should have bee designed to return original future in case of timeout.

Reply:
I'll address it in the followup, I'll be changing the logic here anyway (according to Kamil's yesterday comments).